cpp/src/arrow/util/async_iterator.h

```diff
@@ -0,0 +1,295 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+#include <queue>
+
+#include "arrow/util/functional.h"
+#include "arrow/util/future.h"
+#include "arrow/util/iterator.h"
+#include "arrow/util/optional.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+
+template <typename T>
+using AsyncGenerator = std::function<Future<T>()>;
+
+/// Iterates through a generator of futures, visiting the result of each one and
+/// returning a future that completes when all have been visited
+template <typename T>
+Future<> VisitAsyncGenerator(AsyncGenerator<T> generator,
+                             std::function<Status(T)> visitor) {
```

Comment:
What sort of optimization are you concerned about?  Is it the std::function overhead?  Because I agree, I don't think the compiler is going to optimize that away.  This visitor is probably going to be called on a per-block basis and shouldn't really be used on a per-element basis.  The overhead of calling the function is likely to be on par with the overhead already existing waiting for a future to complete, adding the callback, executing the callback, etc.

Reply:
Yes. All those overheads tend to add up, though. Perhaps this is not overly concerning, but e.g. in the compute layer we're finding out that calling a kernel on relatively small batches has high fixed costs that we didn't envision, simply because of such overheads accumulating.

(I'd also mention that `std::function` makes backtraces in debug mode rather annoying, but that's a much smaller concern)

Reply:
> (I'd also mention that std::function makes backtraces in debug mode rather annoying, but that's a much smaller concern)
I can live with the backtraces (but I agree they are annoying) but it also makes it pretty much impossible to "step into" as well.

Are you sure the overhead is std::function vs function pointer/lambda?  I wonder how much optimization the compiler can really do with an arbitrary callable?

For example, I've been bitten by lambda before when I captured an accumulator variable by reference because the accumulator, which had previously been a register, now required a full RAM write for every increment (since other threads could now conceivably be observing this).

Reply:
Well, at least the templated callable can be inlined, while the `std::function` presumably never will (or at least is much less likely to be inlined).